### PR TITLE
support for closures via opaque closures

### LINKED ISSUE #28: Closures in RuntimeGeneratedFunctions
Just encountered an odd bug testing my [package](https://github.com/0x0f0f0f/Metatheory.jl)

```julia
	Reading Memory: Error During Test at /home/sea/src/julia/Metatheory/test/test_while_interpreter.jl:17
	  Test threw exception
	  Expression: 2 == rewrite(:((x, $(Mem(:x => 2)))), read_mem; order = :inner, m = #= /home/sea/src/julia/Metatheory/test/test_while_interpreter.jl:17 =# @__MODULE__())
	  MethodError: no method matching generated_callfunc(::RuntimeGeneratedFunctions.RuntimeGeneratedFunction{(Symbol("##reducing_expression#257"),), var"#_RGF_ModTag", var"#_RGF_ModTag", (0xa3a3c3bf, 0xb070e435, 0x08893b98, 0x9f9e31fb, 0xf15bd3ff)}, ::Symbol, ::Module)
	  The applicable method may be too new: running in world age 29639, while current world is 29642.
	  Closest candidates are:
	    generated_callfunc(::RuntimeGeneratedFunctions.RuntimeGeneratedFunction{argnames, cache_tag, var"#_RGF_ModTag", id}, ::Any...) where {argnames, cache_tag, id} at none:0 (method too new to be called from this world context.)
	    generated_callfunc(::RuntimeGeneratedFunctions.RuntimeGeneratedFunction{argnames, cache_tag, Metatheory.var"#_RGF_ModTag", id}, ::Any...) where {argnames, cache_tag, id} at none:0
	  Stacktrace:
	    [1] (::RuntimeGeneratedFunctions.RuntimeGeneratedFunction{(Symbol("##reducing_expression#257"),), var"#_RGF_ModTag", var"#_RGF_ModTag", (0xa3a3c3bf, 0xb070e435, 0x08893b98, 0x9f9e31fb, 0xf15bd3ff)})(::Symbol, ::Module)
	      @ RuntimeGeneratedFunctions ~/.julia/packages/RuntimeGeneratedFunctions/tJEmP/src/RuntimeGeneratedFunctions.jl:92
	    [2] (::Metatheory.var"#35#41"{Module})(x::Symbol)
	      @ Metatheory ~/src/julia/Metatheory/src/rewrite.jl:24
	    [3] normalize_nocycle(::Function, ::Symbol; callback::Metatheory.var"#34#40"{Int64})
	      @ Metatheory ~/src/julia/Metatheory/src/util.jl:119
	    [4] #36
	      @ ~/src/julia/Metatheory/src/rewrite.jl:25 [inlined]
	    [5] #df_walk!#6
	      @ ~/src/julia/Metatheory/src/util.jl:30 [inlined]
	    [6] #7
	      @ ~/src/julia/Metatheory/src/util.jl:38 [inlined]
	    [7] |>(x::Symbol, f::Metatheory.var"#7#8"{Vector{Symbol}, Bool, Metatheory.var"#36#42"{Metatheory.var"#35#41"{Module}, Metatheory.var"#34#40"{Int64}}, Tuple{}})
	      @ Base ./operators.jl:859
	    [8] _broadcast_getindex_evalf
	      @ ./broadcast.jl:648 [inlined]
	    [9] _broadcast_getindex
	      @ ./broadcast.jl:621 [inlined]
	   [10] getindex
	      @ ./broadcast.jl:575 [inlined]
	   [11] copy
	      @ ./broadcast.jl:922 [inlined]
	   [12] materialize(bc::Base.Broadcast.Broadcasted{Base.Broadcast.DefaultArrayStyle{1}, Nothing, typeof(|>), Tuple{Vector{Any}, Base.RefValue{Metatheory.var"#7#8"{Vector{Symbol}, Bool, Metatheory.var"#36#42"{Metatheory.var"#35#41"{Module}, Metatheory.var"#34#40"{Int64}}, Tuple{}}}}})
	      @ Base.Broadcast ./broadcast.jl:883
	   [13] df_walk!(::Function, ::Expr; skip::Vector{Symbol}, skip_call::Bool)
	      @ Metatheory ~/src/julia/Metatheory/src/util.jl:38
	   [14] #37
	      @ ~/src/julia/Metatheory/src/rewrite.jl:30 [inlined]
	   [15] (::Metatheory.var"#39#45"{Metatheory.var"#37#43", Metatheory.var"#36#42"{Metatheory.var"#35#41"{Module}, Metatheory.var"#34#40"{Int64}}})(x::Expr)
	      @ Metatheory ~/src/julia/Metatheory/src/rewrite.jl:37
	   [16] normalize_nocycle(::Function, ::Expr; callback::Metatheory.var"#24#26")
	      @ Metatheory ~/src/julia/Metatheory/src/util.jl:119
	   [17] normalize_nocycle(::Function, ::Expr)
	      @ Metatheory ~/src/julia/Metatheory/src/util.jl:117
	   [18] rewrite(ex::Expr, theory::Vector{Rule}; __source__::LineNumberNode, order::Symbol, m::Module, timeout::Int64)
	      @ Metatheory ~/src/julia/Metatheory/src/rewrite.jl:37
	   [19] macro expansion
	      @ ~/src/julia/Metatheory/test/test_while_interpreter.jl:17 [inlined]
	   [20] macro expansion
	      @ ~/src/julia-compiler/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
	   [21] top-level scope
	      @ ~/src/julia/Metatheory/test/test_while_interpreter.jl:17
``` 

How to reproduce: 
Comment lines 7 and 8 in `Metatheory.jl/test/runtests.jl`
https://github.com/0x0f0f0f/Metatheory.jl/blob/a273d9ed6b7d88cadaa3d4f66299de3f3649d719/test/runtests.jl#L7-L8

uncomment last argument (cache module) in line 14 in `Metatheory.jl/test/test_while_interpreter.jl`
https://github.com/0x0f0f0f/Metatheory.jl/blob/a273d9ed6b7d88cadaa3d4f66299de3f3649d719/test/test_while_interpreter.jl#L14

Am I just initializing/using RGF in the wrong way? ([here](https://github.com/0x0f0f0f/Metatheory.jl/blob/a273d9ed6b7d88cadaa3d4f66299de3f3649d719/src/matchcore_compiler.jl#L61) and [also here](https://github.com/0x0f0f0f/Metatheory.jl/blob/a273d9ed6b7d88cadaa3d4f66299de3f3649d719/src/EGraphs/saturation.jl#L36))  I admit that I have done some weird hacks to achieve a behaviour that is partially close to dynamic scoped variable capturing (not technically closures but close to).

### PR DESCRIPTION
**Pull request by simeonschaub**
Should close #28. Kwargs or `where` parameters might not work, but that's an issue in Julia base.

### CODE REVIEW

File: src/RuntimeGeneratedFunctions.jl [hunk 1]
```diff
@@ -44,9 +51,13 @@ end
 """
 struct RuntimeGeneratedFunction{argnames, cache_tag, context_tag, id} <: Function
     body::Expr
-    function RuntimeGeneratedFunction(cache_tag, context_tag, ex)
+    function RuntimeGeneratedFunction(cache_tag, context_tag, ex; opaque_closures=true)
```
Comment (ChrisRackauckas) [inline src/RuntimeGeneratedFunctions.jl:54]: doesn't this need to be set based on the Julia version?

Comment (simeonschaub) [reply src/RuntimeGeneratedFunctions.jl:54]: See the check just below. It's ignored if `Base.Experimental.@opaque` is not defined.

**Comment by ChrisRackauckas** [comment]
Somehow the timing regression tests are failing now?

**Comment by simeonschaub** [comment]
Ah, I just forgot to add the keyword to the other method as well.

**Comment by ChrisRackauckas** [comment]
Great thanks!

**Comment by 0x0f0f0f** [comment]
Thanks!

**Comment by shashi** [comment]
Thanks! This looks great! We needed this pretty badly.

**Comment by c42f** [comment]
Nice! So great that the compiler can now do most of the work for us here :)

**Pull request merged**